debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java

```diff
@@ -297,34 +348,96 @@ protected abstract SchemaChangeEvent getCreateTableEvent(RelationalSnapshotConte
+
+        Map<TableId, String> queryTables = new HashMap<>();
+        Map<TableId, OptionalLong> rowCountTables = new LinkedHashMap<>();
+        for (TableId tableId : snapshotContext.capturedTables) {
+            final Optional<String> selectStatement = determineSnapshotSelect(snapshotContext, tableId);
+            if (selectStatement.isPresent()) {
+                LOGGER.info("For table '{}' using select statement: '{}'", tableId, selectStatement.get());
+                queryTables.put(tableId, selectStatement.get());
+
+                final OptionalLong rowCount = rowCountForTable(tableId);
+                rowCountTables.put(tableId, rowCount);
+            }
+            else {
+                LOGGER.warn("For table '{}' the select statement was not provided, skipping table", tableId);
+                snapshotProgressListener.dataCollectionSnapshotCompleted(snapshotContext.partition, tableId, 0);
             }
+        }
 
-            LOGGER.debug("Snapshotting table {}", tableId);
+        int snapshotOrderByRowCount = connectorConfig.snapshotOrderByRowCount();
+        if (snapshotOrderByRowCount != 0) {
+            LOGGER.info("Sort tables by row count {}", snapshotOrderByRowCount > 0 ? "asc" : "desc");
+            rowCountTables = rowCountTables.entrySet().stream()
+                    .sorted(Map.Entry.comparingByValue((a, b) -> snapshotOrderByRowCount * ((Long) a.orElse(0)).compareTo(b.orElse(0))))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+        }
 
-            createDataEventsForTable(sourceContext, snapshotContext, snapshotReceiver, snapshotContext.tables.forTable(tableId), tableOrder++, tableCount);
+        Queue<O> offsets = new ConcurrentLinkedQueue<O>();
+        offsets.add(snapshotContext.offset);
+        for (int i = 1; i < snapshotMaxThreads; i++) {
+            offsets.add(copyOffset(snapshotContext));
+        }
+
+        try {
+            int tableCount = rowCountTables.size();
+            int tableOrder = 1;
+            for (TableId tableId : rowCountTables.keySet()) {
+                boolean firstTable = tableOrder == 1 && snapshotMaxThreads == 1;
+                boolean lastTable = tableOrder == tableCount && snapshotMaxThreads == 1;
+                String selectStatement = queryTables.get(tableId);
+                OptionalLong rowCount = rowCountTables.get(tableId);
+
+                Callable<Exception> callable = createDataEventsForTableCallable(sourceContext, snapshotContext, snapshotReceiver,
+                        snapshotContext.tables.forTable(tableId), firstTable, lastTable, tableOrder++, tableCount, selectStatement, rowCount, connectionPool, offsets);
+                completionService.submit(callable);
+            }
+
+            Exception firstError = null;
+            for (int i = 0; i < tableCount; i++) {
+                Exception error = completionService.take().get();
```

Comment:
Do we need to return the exception? Can't we use `ExecutionException` mechanism? Also when the first error is hit then all running tasks should be imediatelly cancelled not waiting for their completion.